src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+  KALDI_ASSERT(numDimensions == kConvolutionDimension_);
+  KALDI_ASSERT(mode == CUDNN_CROSS_CORRELATION);
+  KALDI_ASSERT(float_type == cudnn::GetDataType());
+  WriteToken(os, binary, "<FilterXPadding>");
+  WriteBasicType(os, binary, padding[0]);
+  WriteToken(os, binary, "<FilterYPadding>");
+  WriteBasicType(os, binary, padding[1]);
+  WriteToken(os, binary, "<FilterZPadding>");
+  WriteBasicType(os, binary, padding[2]);
+  WriteToken(os, binary, "<FilterXStride>");
+  WriteBasicType(os, binary, strides[0]);
+  WriteToken(os, binary, "<FilterYStride>");
+  WriteBasicType(os, binary, strides[1]);
+  WriteToken(os, binary, "<FilterZStride>");
+  WriteBasicType(os, binary, strides[2]);
+  WriteToken(os, binary, "<FilterXUpscale>");
+  WriteBasicType(os, binary, upscales[0]);
+  WriteToken(os, binary, "<FilterYUpscale>");
+  WriteBasicType(os, binary, upscales[1]);
+  WriteToken(os, binary, "<FilterZUpscale>");
+  WriteBasicType(os, binary, upscales[2]);
+  WriteToken(os, binary, "<InputVectorization>");
+  WriteBasicType(os, binary, static_cast<int32>(input_vectorization_));
+  WriteToken(os, binary, "<FilterParams>");
+  filter_params_.Write(os, binary);
+  WriteToken(os, binary, "<BiasParams>");
+  bias_params_.Write(os, binary);
+  WriteToken(os, binary, "<IsGradient>");
+  WriteBasicType(os, binary, is_gradient_);
+  WriteToken(os, binary, "</CuDNN3DConvolutionComponent>");
+}
+
+void CuDNN3DConvolutionComponent::SetZero(bool treat_as_gradient) {
+  if (treat_as_gradient) {
+    SetActualLearningRate(1.0);
+    is_gradient_ = true;
+  }
+  filter_params_.SetZero();
+  bias_params_.SetZero();
+}
+
+BaseFloat CuDNN3DConvolutionComponent::DotProduct(const UpdatableComponent &other_in) const {
+  const CuDNN3DConvolutionComponent *other =
+    dynamic_cast<const CuDNN3DConvolutionComponent*>(&other_in);
+  return TraceMatMat(filter_params_, other->filter_params_, kTrans) +
+    VecVec(bias_params_, other->bias_params_);
+}
+
+std::string CuDNN3DConvolutionComponent::Info() const {
+  int32 numDimensions;
+  int32 pad_dims[kConvolutionDimension_];
+  int32 stride_dims[kConvolutionDimension_];
+  int32 upscale_dims[kConvolutionDimension_];
+  cudnnConvolutionMode_t mode;
+  cudnnDataType_t float_type;
+  CUDNN_SAFE_CALL(cudnnGetConvolutionNdDescriptor(conv_desc_,
+                                                  kConvolutionDimension_,
+                                                  &numDimensions,
+                                                  pad_dims,
+                                                  stride_dims,
+                                                  upscale_dims,
+                                                  &mode,
+                                                  &float_type)
+                  );
+  KALDI_ASSERT(float_type == cudnn::GetDataType());
+  int32 filter_dims[CUDNN_DIM_MAX];
+  CUDNN_SAFE_CALL(cudnnGetFilterNdDescriptor(filter_desc_,
+                                             kConvolutionDimension_,
```

Comment:
I suspect it to be kConvolutionDimension_+ 2 because at [https://github.com/kaldi-asr/kaldi/pull/770/files#diff-cd5acff10defb49b4be39b32b9c3f6c8R142](url) you passed kConvolutionDimension_+ 2 to cudnnSetFilterNdDescriptor(). Also from the CUDNN manual cudnnGetFilter4dDescriptor() returns number of input/output feature maps along with x/y dim.